addons/crm/models/crm_team_member.py

```diff
@@ -50,163 +68,95 @@ def _constrains_assignment_domain(self):
+                member._get_assignment_quota() > 0
+            ).sorted("lead_day_count")
+            if not team_members:
+                break
+            quota_per_member = {member: member._get_assignment_quota(force_quota=force_quota) for member in team_members}
+            result_data.update({member: {'assigned': self.env['crm.lead'], "quota": quota} for member, quota in quota_per_member.items()})
+            members_dom = {m: literal_eval(m.assignment_domain or '[]') for m in team_members}
+            leads = self.env["crm.lead"].search([
+                    ('user_id', '=', False),
+                    ('date_open', '=', False),
+                    ('team_id', '=', team.id),
+                    ('type', '=', 'lead'),
+                ], order='probability DESC, id')
+            for lead in leads:
+                if not team_members:
+                    break
+                counter += 1
+                for member in team_members:
+                    if not lead.filtered_domain(members_dom[member]):
+                        continue
+                    lead.with_context(mail_auto_subscribe_no_notify=True).convert_opportunity(
+                        lead.partner_id,
+                        user_ids=member.user_id.ids
+                    )
+                    result_data[member]['assigned'] |= lead
+                    team_members -= member
+                    quota_per_member[member] -= 1
+                    if quota_per_member[member] > 0:
+                        # If the member should receive more lead, send him back at the end of the list
+                        team_members |= member
+                    break
+
 
                 if auto_commit and counter % commit_bundle_size == 0:
                     self._cr.commit()
-            else:
-                weights[member_index] = 0
-
-            if weights[member_index] <= 0:
-                population.pop(member_index)
-                weights.pop(member_index)
-
-            # failsafe
-            if counter > 100000:
-                population = list()
 
         if auto_commit:
             self._cr.commit()
-        # log results and return
-        result_data = dict(
-            (member_info["team_member"], {"assigned": member_info["assigned"]})
-            for member_id, member_info in members_data.items()
-        )
-        _logger.info('Assigned %s leads to %s salesmen', len(leads_done_ids), len(members))
+
+        _logger.info('Assigned %s leads to %s salesmen', sum(len(r['assigned']) for r in result_data.values()), len(result_data))
         for member, member_info in result_data.items():
-            _logger.info('-> member %s: assigned %d leads (%s)', member.id, len(member_info["assigned"]), member_info["assigned"])
+            _logger.info('-> member %s of team %s: assigned %d/%d leads (%s)', member.id, member.crm_team_id.id, len(member_info["assigned"]), member_info["quota"], member_info["assigned"])
         return result_data
```

Comment:
As we log above, it's not very useful to return result_data; might be better to not keep such a large set in memory.